### ISSUE TITLE
Results vary greatly across experiments

### ISSUE THREAD

**Issue by William-HYWu** [goombalab/hydra#12]
Dear Authors,
Thanks for your wonderful work! I'm trying to implement hydra on some datasets, but the results vary greatly across identically configured experiments. I've already fixed the random seeds with the following code. Did I miss anything that will lead to different performance with the same configuration?
```
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)

**Comment by William-HYWu** [comment]
The experimental results can shift by about 2% in terms of accuracy

**Comment by sukjunhwang** [comment]
Hi, are you using our BERT training codebase?

**Comment by William-HYWu** [comment]
> Hi, are you using our BERT training codebase?

Hi, no, I'm just using the hydra block.
```
import torch
from .hydra import Hydra

batch, length, dim = 2, 64, 16
x = torch.randn(batch, length, dim).to("cuda")
model = Hydra(
    d_model=dim, # Model dimension d_model
    d_state=64,  # SSM state expansion factor
    d_conv=7,    # Local non-causal convolution width
    expand=2,    # Block expansion factor
    use_mem_eff_path=False,    # Nightly release. Thanks to Alston Lo
).to("cuda")
y = model(x)
assert y.shape == x.shape

**Comment by sukjunhwang** [comment]
Then, the codebase probably needs extra settings for fixing seeds, such as configuring `torch.backends.cudnn.deterministic`

**Comment by William-HYWu** [comment]
> Then, the codebase probably needs extra settings for fixing seeds, such as configuring `torch.backends.cudnn.deterministic`

Thanks for your reply. I think the problem lies in the atomic adds of mamba as mentioned in https://github.com/state-spaces/mamba/issues/137

**Comment by Madou-AI** [comment]
Any news to how to fix the problem that lies in the atomic adds ?

**Comment by William-HYWu** [comment]
> Any news to how to fix the problem that lies in the atomic adds ?

The randomness can be reduced if you use higher precision and a smaller learning rate to train mamba, but personally, I don't know how to address this issue completely.

**Comment by luo-bing** [comment]
> > 关于如何解决原子添加中的问题有什么消息吗？
> 
> 如果你使用更高的精度和更小的学习率来训练 mamba，则可以减少随机性，但就我个人而言，我不知道如何完全解决这个问题。

Hi, have you solved the randomness problem? I'm also encountering this issue and it's very frustrating. I've found that while lowering the learning rate improves stability, the performance also drops significantly.
